data registry views

snopoke opened the pull request:
## Summary
WIP

Mockups: https://www.figma.com/file/vzEYP2maJVa2OzrXMdqomJ/Data-Registries-Mock-ups?node-id=73%3A18

![image](https://user-images.githubusercontent.com/249606/126316073-61b280c0-dc56-45fa-b70a-5183091059b0.png)

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Safety Assurance

- [ ] Risk label is set correctly
- [ ] All migrations are backwards compatible and won't block deploy
- [ ] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [ ] If QA is part of the safety story, the "Awaiting QA" label is used
- [ ] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

### Safety story
<!--
Describe any other pieces to the safety story including
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [ ] This PR can be reverted after deploy with no further considerations

biyeun committed:
remove ENTERPRISE_SSO toggle

joxl committed:
Remove elasticsearch7 dependency and logic

joxl committed:
Make requirements after elasticsearch7 removal

proteusvacuum committed:
Test remote link updates

proteusvacuum committed:
Update remote linked ucr test

biyeun committed:
populate docs url from url helpers

esoergel committed:
Move complexity metrics into python

gherceg committed:
separate data models; separate update methods; add tests

gherceg committed:
handle superuser data models appropriately

gherceg committed:
hide superuser data models in history report

gherceg committed:
update references to ALL_LINKED_MODELS

gherceg committed:
add DomainLinkHistory migration to add previews as choice

gherceg committed:
isort on views.py

gherceg committed:
duplicate toggle events into preview events

gherceg committed:
cleanup tests

sravfeyn committed:
Followup Forms as Case List Form for parent/select modules

sravfeyn committed:
xpath expression to enable case list form

biyeun committed:
ensure that the username is lowercase or else a BadValueError will occur

biyeun committed:
add tests to ensure we handle the case where a username supplied by the IdP has uppercase characters

gherceg committed:
address Jenny's feedback

- remove link to too specific of a line number
- remove NON_SUPERUSER_DATA_MODELS
- add skip on fresh install to migration

sravfeyn committed:
set only if non-empty

sravfeyn committed:
PR feedback

sravfeyn committed:
AdvancedModule doesn't have parent_select

sravfeyn committed:
simplify

sravfeyn committed:
Add test

sravfeyn committed:
Remove label

biyeun committed:
ensure that a new SSO user does not need to go through the domain activation / email confirmation process

We already know their email is valid via the Identity Provider, so we should make sure their first domain is already active.

biyeun committed:
update the new request email to include sso users

biyeun committed:
add tests to ensure intial domains remain active for new SSO users

snopoke committed:
make requirements

sravfeyn committed:
handle other workflow

snopoke committed:
make requirements

snopoke committed:
make requirements

proteusvacuum committed:
Add ability to pull remote reports

proteusvacuum committed:
Ensure domains match

proteusvacuum committed:
Use master id to pull from master domain

proteusvacuum committed:
Show already linked reports

proteusvacuum committed:
Map remote reports too

biyeun committed:
have iframe_close_window inherit from base.html like the other iframe login template

biyeun committed:
add view to display login pending message for SSO users

biyeun committed:
if a SSO user is logging in during session expiration, make sure the login handshake with the Identity Provider happens in a new tab, not in an iframe

biyeun committed:
add a message to local storage after a successful SSO login in a secondary tab so that the parent tab knows when to close the iframe popup

biyeun committed:
this view must be CSRF exempt so that it does not throw a 403 if a user has a second window open with a login pop up and they have already been logged in

biyeun committed:
ensure that superusers logged in with SSO are never given access to superuser functionality

biyeun committed:
rewrite require_superuser decorator to ensure that superusers logged in with SSO never gain access to superuser functionality

biyeun committed:
ensure that users logged in with SSO who have superuser privileges never gain access to admin reports

biyeun committed:
ensure admin tab is never visible to a user logged in with sso, even if they have superuser privileges

sravfeyn committed:
Make select-toggle handle computed options

sravfeyn committed:
remove eof option for followup form

sravfeyn committed:
custom width for media elements

sravfeyn committed:
update text

AmitPhulera committed:
add check for location when no filters are selected

AmitPhulera committed:
remove redundunt check

proteusvacuum committed:
Add data_source property to app summary FormQuestion model

proteusvacuum committed:
Add lookup table source to UI

stephherbers committed:
removed inheritance from other reports templates

stephherbers committed:
added back in js block plus reports/js/filters/main.js that was triggering a js error without it

kaapstorm committed:
Switched unsupported recommonmark for myst-parser

This required upgrading attrs. I checked the changelogs: All looks OK.

kaapstorm committed:
Fix docstring

kaapstorm committed:
Fix docstring

kaapstorm committed:
Fix paths

kaapstorm committed:
Fix reference

kaapstorm committed:
Missing from toctree

stephherbers committed:
removed js block lines used for testing

biyeun committed:
rename iframe_domain_login_new_window to domain_login_new_window

biyeun committed:
ensure that the check for the user redirect on the invitation view matches what is expected for validating the form

biyeun committed:
Update staging.yaml

biyeun committed:
Update staging.yaml

biyeun committed:
update documentation url to live docs

biyeun committed:
Update staging.yaml

biyeun committed:
Update staging.yaml

proteusvacuum committed:
New vars in camel case

proteusvacuum committed:
Add deleted_by param

sravfeyn committed:
update labels conditionally

sravfeyn committed:
lint

andyasne committed:
Disabled Calculate Total from Service Delivery Reports

sravfeyn committed:
trans

sravfeyn committed:
Revert ui change

sravfeyn committed:
check attr

biyeun committed:
Update staging.yaml

biyeun committed:
Update staging.yaml

biyeun committed:
whitespace/indentation fix

biyeun committed:
do not fail hard if request does not have the couch user attribute

biyeun committed:
Update staging.yaml

biyeun committed:
Update staging.yaml

biyeun committed:
Update staging.yaml

biyeun committed:
Update staging.yaml

biyeun committed:
Update staging.yaml

biyeun committed:
lint

joxl committed:
Update translations.

[ci skip]

gherceg committed:
Update staging.yaml

snopoke committed:
add missing method override

snopoke committed:
add test case

snopoke committed:
add toggle

biyeun committed:
Update staging.yaml

orangejenny committed:
Added README.md [ci skip]

gherceg committed:
do not disable downstream flags if disabled upstream

gherceg committed:
do not disable downstream previews if disabled upstream

gherceg committed:
add linked project privilege

biyeun committed:
Update staging.yaml

gherceg committed:
Update staging.yaml

mkangia committed:
don't translate messages

snopoke committed:
registry list static UI

gherceg committed:
used the wrong object manager

stephherbers committed:
make so server options in the dropdown are only for that particular domain

esoergel committed:
make property = '' match both blank and missing values

proteusvacuum committed:
Only show list if data_source is defined

orangejenny committed:
Update staging.yaml

gherceg committed:
unnecessary reference to outdated variable

gherceg committed:
Update staging.yaml

gherceg committed:
Update staging.yaml

shyamkumarlchauhan committed:
Revert "Split feature flags and feature previews into their own data models"

snopoke committed:
reset date modified for legacy sms and email data

stephherbers committed:
Update staging.yaml

snopoke committed:
make query return distinct results

biyeun committed:
Update staging.yaml

gherceg committed:
Revert "Revert "Split feature flags and feature previews into their own data models""

gherceg committed:
Update staging.yaml

gherceg committed:
Update staging.yaml

gherceg committed:
Update staging.yaml

orangejenny committed:
Added TableauView

orangejenny committed:
Update staging.yaml

orangejenny committed:
Removed TableauReport and dynamic classes

orangejenny committed:
Extracted _get_tableau_items

orangejenny committed:
Fixed breadcrumbs and removed TODO

gherceg committed:
remove unused import

im a goofy goober

orangejenny committed:
Updated to 404 instead of 500 when viz is missing

orangejenny committed:
Update corehq/apps/cloudcare/static/cloudcare/js/README.md

Co-authored-by: Ethan Soergel <esoergel@users.noreply.github.com>

orangejenny committed:
Added some content

orangejenny committed:
Replaced cloudcare.rst with new doc

orangejenny committed:
Switched from md to rst

orangejenny committed:
Extracted anatomy section and tried a different table syntax

orangejenny committed:
Fixed sub-lists, perhaps

orangejenny committed:
Switched cells

snopoke committed:
remove 'badge' template

snopoke committed:
add 'new registry' button

snopoke committed:
adjust small screen sizing

snopoke committed:
add JS and convert data binding to knockout

orangejenny committed:
Added more about routing and URLs

orangejenny committed:
Updated ReadTheDocs TOC

orangejenny commented on corehq/apps/registry/static/registry/js/registry_list.js:
```diff
@@ -0,0 +1,109 @@
+            "%(count)s Project Spaces Participating",
+        )
+    }
+    let DeleteModal = function (element) {
```
It's interesting to pass in the element - usually knockout doesn't directly manipulate the DOM in js, other than in custom bindings. The more typical way to do this in our code is to use the `openModal` binding defined in `knockout_bindings.ko.js` and used in places like `releases_table.html`. That said, this code is easy to follow and seems fine.

snopoke replied:
I'll have a look at that

orangejenny commented on corehq/apps/registry/templates/registry/registry_list.html:
```diff
@@ -0,0 +1,169 @@
+  <div id="data-registry-list">
+    <div class="row">
+      <div class="col-sm-8 col-md-6">
+        <h3>Data Registries</h3>
```
trans

orangejenny commented on corehq/apps/registry/templates/registry/registry_list.html:
```diff
@@ -0,0 +1,169 @@
+        <h3>Data Registries</h3>
+      </div>
+      <div class="col-sm-4 col-md-3">
+        <div class="pull-right" style="margin-top: 17px; margin-bottom: 8.5px">
```
Nitpick: I'm guessing the inline styles are to keep the button vertically aligned with the header - would it work if you put this div/button inside of the h3 instead of using the row & cols?

snopoke replied:
I think I tried that but I'll give it another go

orangejenny commented on corehq/apps/registry/templates/registry/registry_list.html:
```diff
@@ -0,0 +1,169 @@
+      <div class="col-sm-12 col-md-9">
+      <!-- ko foreach: ownedRegistries -->
+      <div class="panel panel-default show-focus">
+        <div class="panel-heading clickable">
```
What happens when you click on the header, are these collapsible?

snopoke replied:
It will go to the edit page - should I add a pencil icon or something?

orangejenny replied:
How about making the title text into a link, with or without a pencil? I don't think people will expect to change pages when they click a panel header, a link is a better signal that you're going to leave the page.

orangejenny commented on corehq/apps/hqwebapp/static/registry/less/registry.less:
```diff
@@ -0,0 +1,19 @@
+
+@import '../../hqwebapp/less/_hq/includes/variables.less';
+
+.label-plain {
```
Curious if you tried out bootstrap's `label-default`, which is pretty similar, although the text is smaller.

snopoke replied:
I did try that - it has a very dark background:

![image](https://user-images.githubusercontent.com/249606/126461727-1deece2c-7dea-45ce-ae4a-3d597d43dcf2.png)
![image](https://user-images.githubusercontent.com/249606/126461521-aec9c4ae-b350-4e52-bd97-8322420eea95.png)

orangejenny replied:
hrm, yeah, that's uglier, and the spacing is weird. The thing that's bugging me a little about the new style is that it looks like a button, with the colored background + border, I think people are going to click on it. What if it just didn't have the background?

orangejenny commented on corehq/apps/registry/templates/registry/registry_list.html:
```diff
@@ -0,0 +1,169 @@
+        </div>
+        <div class="panel-body">
+          <p data-bind="text: description"></p>
+          <p><small><b>{% trans "Owner" %}:</b><b data-bind="text: domain_name"></b></small></p>
```
Nitpick: `<p class="help-block">` would be a similar effect and wouldn't use the non-semantic small and b tags.

orangejenny commented on corehq/apps/registry/templates/registry/registry_list.html:
```diff
@@ -0,0 +1,169 @@
+          {% endblocktrans %}
+        </p>
+        <ul>
+          <li>Applications using the registry for case search</li>
```
trans for this list and the buttons below

orangejenny reviewed (commented):
looks good so far

gherceg committed:
remove unlinking cleanup

gherceg committed:
Update staging.yaml

gherceg committed:
Update staging.yaml

mkangia committed:
specify to store only raw messages

gherceg committed:
add release management priv to enterprise_plan_v1

orangejenny committed:
Update staging.yaml

Robert-Costello committed:
Update staging.yaml

gherceg committed:
update text for release management role

orangejenny committed:
Update staging.yaml

stephherbers committed:
Update staging.yaml

stephherbers committed:
Update staging.yaml

minhaminha committed:
Update staging.yaml

esoergel committed:
Fix build

capture output isn't supported until 3.7

esoergel committed:
Run static analysis build on every PR

This won't send to datadog, but the output will be available in the
travis output, and this will prevent bugs from being introduced without
notice

esoergel committed:
Misc cleanup

gherceg committed:
Update staging.yaml

snopoke committed:
space

snopoke committed:
add es6 support to eslintrc

snopoke commented on corehq/apps/registry/static/registry/js/registry_list.js:
```diff
@@ -0,0 +1,109 @@
+    initialPageData
+) {
+    let getBadgeText = function (count, singular, plural) {
+        return interpolate(ngettext(singular, plural, count), {"count": count}, true);
```
good question, I'll check

snopoke replied:
The answer is no, they won't get picked up like this

snopoke committed:
div to span for labels

snopoke committed:
translate strings

snopoke committed:
nest button in H3 for alignment

snopoke committed:
use knockout modal binding

snopoke committed:
use help-block

snopoke committed:
fix JS translations

snopoke committed:
drop delete button on list + less intrusive labels

snopoke committed:
fix rejected text

snopoke committed:
move translation function to new module

snopoke committed:
convert to observables

snopoke committed:
accept invitation

snopoke committed:
add modal for accept

snopoke committed:
add reject and move actions to new JS module

snopoke committed:
cleanup + remove invite button from list
